packages/react-sdk/src/client.ts

```diff
@@ -0,0 +1,595 @@
+      this.user.attributes = userInfo.attributes
+    }
+    if (!this.isUserPromiseResolved) {
+      this.userPromiseResovler(this.user)
+      this.isUserPromiseResolved = true
+    }
+    this.onUserUpdateHandlers.forEach(handler => handler(this.user))
+  }
+
+  onUserUpdate(handler: OnUserUpdateHandler): DisposeFn {
+    this.onUserUpdateHandlers.push(handler)
+
+    return () => {
+      const ind = this.onUserUpdateHandlers.indexOf(handler)
+      if (ind > -1) {
+        this.onUserUpdateHandlers.splice(ind, 1)
+      }
+    }
+  }
+
+  /**
+   * @param {string} experimentKey
+   * @param {string} userId
+   * @param {UserAttributes} [attributes]
+   * @returns {(string | null)}
+   * @memberof OptimizelySDKWrapper
+   */
+  public activate(
+    experimentKey: string,
+    overrideUserId?: string,
+    overrideAttributes?: UserAttributes,
+  ): string | null {
+    const [userId, userAttributes] = this.getUserIdAndAttributes(
+      overrideUserId,
+      overrideAttributes,
+    )
+    if (userId === null) {
+      logger.info(
+        'Not activating experiment "%s" because userId is not set',
+        experimentKey,
+      )
+      return null
+    }
+    return this.client.activate(experimentKey, userId, userAttributes)
+  }
+
+  /**
+   *
+   *
+   * @param {string} experimentKey
+   * @param {string} userId
+   * @param {UserAttributes} [attributes]
+   * @returns {(string | null)}
+   * @memberof OptimizelySDKWrapper
+   */
+  public getVariation(
+    experimentKey: string,
+    overrideUserId?: string,
+    overrideAttributes?: UserAttributes,
+  ): string | null {
+    const [userId, userAttributes] = this.getUserIdAndAttributes(
+      overrideUserId,
+      overrideAttributes,
+    )
+    if (userId === null) {
+      logger.info(
+        'getVariation returned null for experiment "%s" because userId is not set',
+        experimentKey,
+      )
+      return null
+    }
+    return this.client.getVariation(experimentKey, userId, userAttributes)
+  }
+
+  track(
+    eventKey: string,
+    overrideUserId?: string | EventTags,
+    overrideAttributes?: UserAttributes,
+    eventTags?: EventTags,
+  ) {
+    if (typeof overrideUserId !== 'undefined' && typeof overrideUserId !== 'string') {
+      eventTags = overrideUserId
+      overrideUserId = undefined
+      overrideAttributes = undefined
+    }
+    const [userId, attributes] = this.getUserIdAndAttributes(
+      overrideUserId,
+      overrideAttributes,
+    )
+
+    if (userId === null) {
+      logger.info(
+        'track for event "%s" not being sent because userId is not set',
+        eventKey,
+      )
+      // TODO log
+      return
+    }
+
+    return this.client.track(eventKey, userId, attributes, eventTags)
+  }
+
+  /**
+   * Note: in the case where the feature isnt in the datafile or the datafile hasnt been
+   * loaded, this will return `false`
+   *
+   * @param {string} feature
+   * @param {string} userId
+   * @param {UserAttributes} [attributes]
```

Comment:
is `attributes` supposed to be enclosed in `[ ]`?

Reply:
Yea - this means optional parameter in JSDoc.